crates/notedeck/src/submgr.rs

```diff
@@ -0,0 +1,269 @@
+}
+
+#[derive(Debug, Default)]
+pub struct SubSpecBuilder {
+    rmtid: Option<String>,
+    filters: Vec<Filter>,
+    constraints: Vec<SubConstraint>,
+}
+
+impl SubSpecBuilder {
+    pub fn new() -> Self {
+        SubSpecBuilder::default()
+    }
+    pub fn rmtid(mut self, id: String) -> Self {
+        self.rmtid = Some(id);
+        self
+    }
+    pub fn filters(mut self, filters: Vec<Filter>) -> Self {
+        self.filters.extend(filters);
+        self
+    }
+    pub fn constraint(mut self, constraint: SubConstraint) -> Self {
+        self.constraints.push(constraint);
+        self
+    }
+    pub fn build(self) -> SubSpec {
+        let mut outbox_relays = Vec::new();
+        let mut allowed_relays = Vec::new();
+        let mut blocked_relays = Vec::new();
+        let mut is_oneshot = false;
+        let mut is_local = false;
+
+        for constraint in self.constraints {
+            match constraint {
+                SubConstraint::OneShot => is_oneshot = true,
+                SubConstraint::Local => is_local = true,
+                SubConstraint::OutboxRelays(relays) => outbox_relays.extend(relays),
+                SubConstraint::AllowedRelays(relays) => allowed_relays.extend(relays),
+                SubConstraint::BlockedRelays(relays) => blocked_relays.extend(relays),
+            }
+        }
+
+        SubSpec {
+            rmtid: self.rmtid,
+            filters: self.filters,
+            outbox_relays,
+            allowed_relays,
+            blocked_relays,
+            is_oneshot,
+            is_local,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct SubSpec {
+    rmtid: Option<String>,
+    filters: Vec<Filter>,
+    outbox_relays: Vec<String>,
+    allowed_relays: Vec<String>,
+    blocked_relays: Vec<String>,
+    is_oneshot: bool,
+    is_local: bool,
+}
+
+pub struct SubMgr<'a> {
+    ndb: &'a mut Ndb,
+    subs: BTreeMap<SubId, SubSpec>,
+}
+
+impl<'a> SubMgr<'a> {
+    pub fn new(ndb: &'a mut Ndb) -> Self {
+        SubMgr {
+            ndb,
+            subs: BTreeMap::new(),
+        }
+    }
+
+    pub fn subscribe(&mut self, spec: SubSpec) -> SubResult<SubReceiver> {
+        let receiver = self.make_subscription(&spec)?;
+        self.subs.insert(receiver.id, spec);
+        Ok(receiver)
+    }
+
+    pub fn unsubscribe(&mut self, rcvr: &SubReceiver) -> SubResult<()> {
+        self.subs.remove(&rcvr.id);
+        Ok(())
+    }
+
+    fn make_subscription(&mut self, sub: &SubSpec) -> SubResult<SubReceiver> {
+        let subscription = self.ndb.subscribe(&sub.filters)?;
+        let mut stream = subscription.stream(self.ndb).notes_per_await(1);
+        Ok(SubReceiver::new(subscription.into(), stream))
+    }
+
+    // Unfortunately this is needed so the caller doesn't need
+    // concurrent mutable and immutable references to ndb.  I think
+    // making process_event use a mutable reference to ndb might be a
+    // more obvious fix
```

Comment:
sounds like this shouldn't be an issue at all.

Reply:
Fix above fixed it ...